components/optimizer/base.py

```diff
@@ -0,0 +1,40 @@
+import abc
+from common.minimodel import Group, Plan
+from components.selector import GroupInfo
+from typing import Dict
+
+Selection = Dict[Group, GroupInfo]
+
+
+class BaseOptimizer(metaclass=abc.ABCMeta):
+    """
+    Base class for all Optimizer components.
+    Each optimizing algorithm needs to implement the following methods:
+
+    schedule: methods that triggers the formation of the plan
+    get_visits: methods that updates the plan 
+
+    """
+
+    @classmethod
+    def __subclasshook__(cls, subclass):
+        return (hasattr(subclass, 'schedule') and
+                callable(subclass.schedule) and
+                hasattr(subclass, 'add') and
+                callable(subclass.add) and
+                hasattr(subclass, 'get_visits') and
+                callable(subclass.get_visits) and
+                hasattr(subclass, '_run') and
+                callable(subclass, '_run'))
```

Comment:
And maybe this is something that you know that I don't, because I haven't worked with `__subclasshook__` before and had to look it up, but why don't we just use `@abstractmethod`s for these? Wouldn't we want them all to have the same type, and have the type specified?